README
This is Perl module HTTP::Engine::Middleware.

INSTALLATION

HTTP::Engine::Middleware installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan HTTP::Engine::Middleware

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

HTTP::Engine::Middleware documentation is available as in POD. So you can do:

    % perldoc HTTP::Engine::Middleware

to read the documentation online with your favorite pager.

Tokuhiro Matsuno


Changes
Revision history for Perl extension HTTP::Engine::Middleware

0.22    2010-04-13T12:58:26+09:00
        - ReverseProxy
          case insensitive for HTTPS headers (x-forwarded-https, x-forwarded-proto)

0.21    2010-02-18T17:22:49+09:00
        - Static: Fixed crashing when rendering file with unknown extension

0.20    2010-02-18T17:08:49+09:00
        - MobileAttribute: fixed uncleanup request object

0.19    2010-01-29T19:35:32+09:00
        - UploadTemp is deprecated

0.18    2010-01-04T12:38:16+09:00
        - ReverseProxy
          use the last item of x-forwarded-host as host

0.17    2009-07-06T18:57:19+09:00
        - ReverseProxy
          !!!!! CAUTIONS !!!!! CAUTIONS !!!!! CAUTIONS !!!!! CAUTIONS !!!!! CAUTIONS !!!!!
          It stopped working from other than a specific crient IP address.
          Please set up the new allowed_remote option for ReverseProxy to work in your network.
          see L<HTTP::Engine::Middleware::ReverseProxy>
          !!!!! CAUTIONS !!!!! CAUTIONS !!!!! CAUTIONS !!!!! CAUTIONS !!!!! CAUTIONS !!!!!

0.16    2009-07-02T20:00:51+09:00
        - DoCoMoGUID:
          fixed tests # RT #47477 (tokuhirom)

0.15    2009-06-19T20:12:49+09:00
        - core
          latest Mouse has a compatible interface with Moose (tokuhirom)
        - DoCoMoGUID:
          change dependency to HTML::StickyQuery::DoCoMoGUID

0.14    2009-06-04T13:16:16+09:00
        - Static: fixed a bug around is_404_handler:
                  403 status returned when access came to the directory doesn't exist.
                  It should be handled by 404 handler.

0.13    2009-05-21T17:02:49+09:00
        - We moved to GitHub ( http://github.com/http-engine )
        - Static: add is_404_handler option
                  fixed Last-Modified send to epoch time bug
        - DebugScreen: bug fix should not catch the exception in eval-block
        - depend to Any::Moose 0.08 ( change moose loaded check function s/is_moose_loaded/moose_is_preferred/ )
        - fixed to bug which does not work when there is no Response object
          DoCoMoGUID, Encode, FillInForm, HTTPSession, Profile::Runtime

0.12    2009-05-01T17:57:20+09:00
        - add Middleware::AccessLog  # tokuhirom++
        - add Middleware::UploadTemp # upload files cleanup manager

0.11    2009-04-10T19:03:24+09:00
        - DebugScreen: utf8 downgread to response body

0.11_01 2009-04-07T13:41:05+09:00
        - refactoring some code to HTTP/Engine/Middleware.pm
        - remove (inner|outer)_middleware support. this feature is perfectly broken.
        - support multiple instance, exactly

	- added Middleware::Status
          and Middleware::Status::Memory
        - Middleware::Static mime_type is manually stringify

0.10    2009-02-27T11:46:35+09:00
        - Encode: fixed some problem on charset of content-type
        - HTTPSession: fixed create instance duped

0.09    2009-02-24T15:08:00+09:00
        - fixed moose's test doesnt work bug (required MooseX::Types)
        - fixed HTTPSession's test doesnt work bug (required HTML::StickyQuery)
        - forget use HTTP::Session in Middleware::HTTPSession

0.08    2009-02-20T14:27:42+09:00
         - release to Any::Moose HTTP::Engine::Middleware

0.08_02 2009-02-19T12:20:07+09:00
        - depend explicitly for Any::Moose and Mouse

0.08_01 2009-02-18T22:40:51+09:00
        - switched to Any::Moose

0.07    2009-02-18T20:54:40+09:00
        - added HTTPSession # tokuhirom++

0.06    2009-02-17T13:32:06+09:00
        - fixed to Middleware::MobileAttributes's test bug

0.05    2009-02-16T18:12:40+09:00
        - dependent to perl 5.8
        - Middleware::MobileAttribute
          pluggable injection style
        - Middleware::Static
          pod fix ( NAME section )
        - fixed to profile runtime test's bug ( last test case )
        - tidy up some middleware's test code

0.04    2009-02-15T14:29:01+09:00
        - few fix release
        - Middleware::Static: yaml friendly configuration
        - Middleware::DebugRequest: fixed bug to test file

0.03    2009-02-14T15:31:15+09:00
        - added simple docs for each middlewares
        - refactoring to log API for middleware
        - Middleware::Static
          change design. new one is simple
          add customizable DirectoryIndex
          fixed vulnerability: directory traversal
          fixed bug to config breaks in Interface::Test and Interface::ModPerl
        - tokurhiom++

0.02    2009-02-12T23:08:46+09:00
        - add DebugRequest, FillInForm, Static
        - Unicode merged to Encode
        - Core: fixed to exception handling bug
        - ReverseProxy: fixed host/port detection bug
        - fixed to profile runtime test's bug, use Scalar::Util::looks_like_number

0.01    2009-02-10T15:58:56+09:00
        - original version


Makefile.PL
use inc::Module::Install;
BEGIN
{
    if (!$ENV{SHIPIT} && eval {
        require Moose  && Moose->VERSION(
            '0.73'
        ) && require MooseX::Types && MooseX::Types->VERSION(
            '0.12'
        )
    }) {
        *HAVE_MOOSE = sub { 1 };
    } else {
        *HAVE_MOOSE = sub { 0 };
    }
}

{
    no warnings 'redefine';
    *recommends_hack = $Module::Install::VERSION > 0.70 ?
        sub { recommends(@_); $_[0] } :
        \&recommends;
}

name 'HTTP-Engine-Middleware';
all_from 'lib/HTTP/Engine/Middleware.pm';

requires 'Any::Moose'   => '0.09';
requires 'Mouse'        => '0.23';
requires 'HTTP::Engine' => '0.03004';
requires 'HTTP::Request';

features(

# For Developer's

    'ModuleReload' => [
        -default => 0,
        recommends_hack('Module::Reload'),
    ],

    'DebugScreen' => [
        -default => 0,
        recommends_hack('CGI::ExceptionManager'),
        recommends_hack('Scope::Upper'),
    ],

    'DebugRequest' => [
        -default => 0,
        recommends_hack('Text::SimpleTable'),
    ],

    'Status::Memory' => [
        -default => 0,
        recommends_hack('B::TerseSize'),
        recommends_hack('Devel::Symdump'),
    ],

    'AccessLog' => [
        -default => 0,
        recommends_hack('DateTime'),
    ],

# static file maneger ( developer ? )

    'Static' => [
        -default => 0,
        recommends_hack('MIME::Types'),
        recommends_hack('Path::Class'),
        recommends_hack('MouseX::Types::Path::Class' => '0.05'),
        recommends_hack('HTTP::Date'),
    ],

# fill in form

    'FillInForm' => [
        -default => 0,
        recommends_hack('HTML::FillInForm' => '2.00'),
    ],

# encoding

    'Encode' => [
        -default => 0,
        recommends_hack('Data::Visitor' => '0.23'),
        recommends_hack('Data::Visitor::Encode'),
        recommends_hack('Encode'),
    ],

# For Japanese Mobile

    'MobileAttribute' => [
        -default => 0,
        recommends_hack('HTTP::MobileAttribute'),
    ],

    'DoCoMoGUID' => [
        -default => 0,
        recommends_hack('HTML::StickyQuery::DoCoMoGUID'),
    ],

# session support

    'HTTPSession' => [
        -default => 0,
        recommends_hack('MouseX::Types'),
        recommends_hack('HTTP::Session'),
    ],
);

my @tests = qw( t/*.t t/*/*.t t/*/*/*.t );
author_tests 'xt';

if (HAVE_MOOSE) {
    my $tests = 
        join ' ', (
            @tests,
            (map {
                my $t = $_;
                $t =~ s{^t/}{t/moose/};
                ($t eq 't/moose/*.t' || $t eq 't/moose/*/*.t') ? () : $t;
            } @tests)
        );
    tests($tests);
} else {
    tests(join ' ', @tests);
}

build_requires 'IO::Scalar';
build_requires 'Test::More';
build_requires 'YAML';
auto_set_repository;
use_test_base;
auto_include;
WriteAll;

# generate moose's test
if (HAVE_MOOSE) {
    warn "generating to Moose's test... please wait few minutes";
    my @testfiles;
    require File::Find;
    require File::Path;
    require File::Spec;

    File::Path::rmtree(File::Spec->catfile( 't', 'moose' ));

    File::Find::find(
        sub {
            return if $File::Find::dir =~ /\./ || /^\./; # skip dot directory
            return unless -f $_;

            my @dirs = File::Spec->splitpath( $File::Find::dir );
            while (!!!$dirs[0]) {
                shift @dirs;
            }
            shift @dirs;
            return if $dirs[0] eq 'xxx'; # skip to moose's test directory remover
            push @testfiles, [ $File::Find::dir, File::Spec->catfile( 't', 'moose', @dirs ), $_ ];
        }, 't'
    );

    for my $stuff (@testfiles) {
        File::Path::mkpath($stuff->[1]);

        my $from = File::Spec->catfile($stuff->[0], $stuff->[2]);
        my $to   = File::Spec->catfile($stuff->[1], $stuff->[2]);

        open my $from_fh, '<', $from;
        open my $to_fh, '>', $to;
        if ($to =~ /\.t$/) {
            print $to_fh "BEGIN { \$ENV{ANY_MOOSE} = 'Moose'; }\n";
        }

        while (<$from_fh>) {
            print $to_fh $_;
        }
    }
    warn "... done";
}
